### PR TITLE
Week 4: Map and Translate

### PR DESCRIPTION
**Pull request by hollyyuqizheng**
Map feature [link](https://github.com/googleinterns/step/blob/master/walkthroughs/week-4-libraries/maps/maps-walkthrough.md)
- The map displays 3 premade markers that I have added myself. 
- Users can click on the map and add new markers with description.
- The newly added marker’s latitude and longitude will also be displayed in a list below the map.

Translation feature [link](https://github.com/googleinterns/step/blob/master/walkthroughs/week-4-libraries/translation/translation-walkthrough.md)
- Proof-of-concept implementation of translation feature: a button triggers API call to translate the image caption. 

Notes: The "quotes.js" file content was extracted from original "script.js". I reorganized the old script.js and divided the code into separate files (quotes.js, map.js, translate.js) according to their purpose.

### CODE REVIEW

File: portfolio/src/main/webapp/map.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+var map;
+
+var allMarkers; 
+
+/** Creates a map and adds it to the page. */
+function createMap() {
+  allMarkers = [];
+
+  // Center the map at the middle of the Pacific so that both sides can show. 
+  const latlongMidwayIsland = {lat: 28.2072, lng: -177.3735};
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/map.js]: Can we follow camel-case by changing latlong to latLong? Here and below.

Comment (jamesnam-google) [reply portfolio/src/main/webapp/map.js]: Also, consider using consistent terminology between long and lng. Maybe we can rename it to,

latLngMidwayIsland

File: portfolio/src/main/webapp/map.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    });
+
+  marker.setMap(map);
+  allMarkers.push(marker); 
+  createUserInfoWindow(map, marker, markerJson.content); 
+}
+
+/** Creates the info window for a marker created by the user. */
+function createUserInfoWindow (map, marker, content) {
+  const infowindow = new google.maps.InfoWindow({
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/map.js]: infowindow => infoWindow

File: portfolio/src/main/webapp/map.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+function postMarker(markerJson) { 
+  const params = new URLSearchParams();
+  params.append('lat', markerJson.lat);
+  params.append('lng', markerJson.lng);
+  params.append('content', markerJson.content);
+  fetch('/marker', {method: 'POST', body: params})
+      .then(editMarker.setMap(null)); 
+}
+
+/** Adds a new marker's latitude and longitude information to the
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/map.js]: Adds can start in the next line.

ex:
/**
 * Adds...

File: portfolio/src/main/java/com/google/sps/servlets/DeleteMarkerServlet.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    PreparedQuery markers = datastore.prepare(query);
+
+    List<Key> toDeleteKeyList = new ArrayList<Key>();
+    markers.asIterable().forEach((markerEntity) -> {
+      toDeleteKeyList.add(markerEntity.getKey());
+    }); 
+
+    datastore.delete(toDeleteKeyList);
+  }
+}
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteMarkerServlet.java]: Let's have a new line at the end

File: portfolio/src/main/java/com/google/sps/data/Marker.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+package com.google.sps.data;
+
+/** Represents a marker on the map. */
+public class Marker {
+
+  public final long id; 
+  public final double lat;
+  public final double lng;
```
Comment (jamesnam-google) [inline portfolio/src/main/java/com/google/sps/data/Marker.java:22]: I noticed that sometimes you're using lng and sometimes you're using long. Please consider using consistent terminology. I think we can use lng (and Lng) consistently in names.

File: portfolio/src/main/webapp/map.js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    createMarkerForEdit(event.latLng.lat(), event.latLng.lng());
+  });
+}
+
+/**
+ * Creates a list of already-set markers for the map. 
+ * These markers cannot be deleted by the user. 
+ */
+function createPremadeMarkers() {
+  const latlongBrownUniv = {lat: 41.8268, lng: -71.4025};
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/map.js]: As I suggested above, consider renaming to

latLngBrownUniv

File: portfolio/src/main/webapp/translate.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+  resultContainer.innerText = 'Loading...';
+
+  const params = new URLSearchParams();
+  params.append('elementId', elementId); 
+  params.append('textToTranslate', textToTranslate);
+  params.append('languageCode', languageCode);
+
+  fetch('/translate', {method: 'POST',body: params})
+      .then(response => response.text())
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/translate.js]: I think this line is indented excessively. We can do,

fetch(...)
  .then(...)

Reference:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

Comment (jamesnam-google) [reply portfolio/src/main/webapp/translate.js]: (two spaces)

File: portfolio/src/main/webapp/staticContent.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  const hobbyContainer = document.getElementById('hobby-container');
+  hobbyContainer.innerText = hobbies;
+  
+  const hobbyWrapper = document.getElementById('hobby-wrapper');
+  hobbyWrapper.style.backgroundColor = 'antiquewhite';
+}
+
+/**
+ * Adds description for each project in project section. 
+ * Input: string -- name of the selected project
```
Comment (jamesnam-google) [inline portfolio/src/main/webapp/staticContent.js]: Consider using @param and @return

Reference:
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

**Review by jamesnam-google — APPROVED** [review]
Approving the change with comments.

File: portfolio/src/main/webapp/map.js (hunk 1 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/map.js]: Updated.

File: portfolio/src/main/webapp/map.js (hunk 2 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/map.js]: Updated.

File: portfolio/src/main/webapp/map.js (hunk 3 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/map.js]: Updated.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteMarkerServlet.java (hunk 4 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteMarkerServlet.java]: Updated.

File: portfolio/src/main/webapp/map.js (hunk 6 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/map.js]: Updated.

File: portfolio/src/main/java/com/google/sps/data/Marker.java (hunk 5 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/java/com/google/sps/data/Marker.java:22]: I see. Updated accordingly.

File: portfolio/src/main/webapp/staticContent.js (hunk 8 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/staticContent.js]: Updated.

File: portfolio/src/main/webapp/translate.js (hunk 7 above)
Comment (hollyyuqizheng) [reply portfolio/src/main/webapp/translate.js]: Updated.

**Review by rgcase — APPROVED** [review]

**Pull request merged**